### PR TITLE
feat(vector): Add optional Ingress resource

### LINKED ISSUE #52: Add ingress to vector chart
Needed to create an AWS ALB

ex:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: vector
  annotations:
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/healthcheck-port: '8686'
    alb.ingress.kubernetes.io/healthcheck-path: /health
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 8080}]'
    alb.ingress.kubernetes.io/target-type: ip
    kubernetes.io/ingress.class: alb
spec:
  rules:
    - host: vector.demo.com
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: vector
                port:
                  name: datadog-agents
```

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

Closes #52

### CODE REVIEW

File: charts/vector/templates/ingress.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
+{{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" -}}
+apiVersion: networking.k8s.io/v1
+{{- else if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1" -}}
+apiVersion: networking.k8s.io/v1beta1
+{{- else -}}
+apiVersion: extensions/v1beta1
+{{- end }}
```
Comment (spencergilbert) [inline charts/vector/templates/ingress.yaml:13]: Ensuring we create the correct resource version depending on what's available/how old the target cluster is

File: charts/vector/templates/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{{- if .Values.ingress.enabled -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
+  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
```
Comment (spencergilbert) [inline charts/vector/templates/ingress.yaml:6]: Ensuring we set the ingressClass properly depending on the available resource version

File: charts/vector/templates/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        paths:
+          {{- range .paths }}
+          - path: {{ .path }}
+            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
+            pathType: {{ .pathType }}
+            {{- end }}
+            backend:
+              {{- if semverCompare ">=1.19-0" $.Capabilities.KubeVersion.GitVersion }}
+              service:
+                name: {{ if $.Values.haproxy.enabled }}{{ include "haproxy.fullname" $ }}{{ else }}{{ include "vector.fullname" $ }}{{ end }}
```
Comment (spencergilbert) [inline charts/vector/templates/ingress.yaml:50]: This and ln 59 toggles the backend service between the vector and vector-haproxy depending on if the haproxy is enabled, saving us from having to have a second ingress resource specifically for the haproxy.

File: charts/vector/templates/ingress.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                port:
+                  {{- with .port.name }}
+                  name: {{ . }}
+                  {{- end }}
+                  {{- with .port.number }}
+                  number: {{ . }}
+                  {{- end }}
```
Comment (spencergilbert) [inline charts/vector/templates/ingress.yaml:57]: Newer ingress versions can specify either a named port or the number, using both is invalid IIRC - I could add an `if/else` here to ensure we don't create an invalid resource... I guess I should just do that...

File: charts/vector/templates/ingress.yaml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+{{- if .Values.ingress.enabled -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
```
Comment (jszwedko) [inline charts/vector/templates/ingress.yaml:2]: Nifty

**Review by jszwedko — APPROVED** [review]
Nice!

File: charts/vector/templates/ingress.yaml (hunk 5 above)
Comment (spencergilbert) [reply charts/vector/templates/ingress.yaml:2]: Shamelessly stolen from the "default" resource that's created from `helm create <chart>`

**Comment by jszwedko** [comment]
The PR comments were helpful, thanks @spencergilbert !

File: charts/vector/templates/ingress.yaml (hunk 1 above)
Comment (spencergilbert) [reply charts/vector/templates/ingress.yaml]: This _didn't_ work and I was being too clever, used the more common `semverCompare here instead

**Pull request merged**